command/cmd_copy.go

```diff
@@ -0,0 +1,632 @@
+	if err != nil {
+		return err
+	}
+
+	metadata := map[string]string{
+		"StorageClass": string(storageClass),
+	}
+
+	err = shouldOverride(dst)
+	if err != nil {
+		if isWarning(err) {
+			printDebug(op, src, dst, err)
+			return nil
+		}
+		return err
+	}
+
+	err = srcClient.Copy(
+		ctx,
+		src,
+		dst,
+		metadata,
+	)
+	if err != nil {
+		return err
+	}
+
+	if deleteSource {
+		if err := srcClient.Delete(ctx, src); err != nil {
+			return err
+		}
+	}
+
+	msg := log.InfoMessage{
+		Operation:   op,
+		Source:      src,
+		Destination: dst,
+		Object: &storage.Object{
+			URL:          dst,
+			StorageClass: storage.StorageClass(storageClass),
+		},
+	}
+	log.Info(msg)
+
+	return nil
+}
+
+func guessContentType(rs io.ReadSeeker) string {
+	defer rs.Seek(0, io.SeekStart)
+
+	const bufsize = 512
+	buf, err := ioutil.ReadAll(io.LimitReader(rs, bufsize))
+	if err != nil {
+		return ""
+	}
+
+	return http.DetectContentType(buf)
+}
+
+func givenCommand(c *cli.Context) string {
+	return fmt.Sprintf("%v %v", c.Command.FullName(), strings.Join(c.Args().Slice(), " "))
+}
+
+// prepareCopyDestination will return a new destination URL for local->local
+// and remote->remote copy operations.
+func prepareCopyDestination(
+	ctx context.Context,
+	originalSrc *objurl.ObjectURL,
+	src *objurl.ObjectURL,
+	dst *objurl.ObjectURL,
+	parents bool,
+) (*objurl.ObjectURL, error) {
+	objname := src.Base()
+	if parents {
+		objname = src.Relative()
+	}
+
+	// For remote->remote copy operations, treat <dst> as prefix if it has "/"
+	// suffix.
+	if dst.IsRemote() {
+		if strings.HasSuffix(dst.Path, "/") {
+			dst = dst.Join(objname)
+		}
+		return dst, nil
+	}
+
+	client, err := storage.NewClient(dst)
+	if err != nil {
+		return nil, err
+	}
+
+	// For local->local copy operations, we can safely stat <dst> to check if
+	// it is a file or a directory.
+	obj, err := client.Stat(ctx, dst)
+	if err != nil && err != storage.ErrGivenObjectNotFound {
+		return nil, err
+	}
+
+	// Absolute <src> path is given. Use given <dst> and local copy operation
+	// will create missing directories if <dst> has one.
+	if !originalSrc.HasGlob() {
+		return dst, nil
+	}
+
+	// For local->local copy operations, if <src> has glob, <dst> is expected
+	// to be a directory. As always, local copy operation will create missing
+	// directories if <dst> has one.
+	if obj != nil && !obj.Type.IsDir() {
+		return nil, fmt.Errorf("destination argument is expected to be a directory")
+	}
+
+	return dst.Join(objname), nil
+}
+
+// prepareDownloadDestination will return a new destination URL for
+// remote->local and remote->remote copy operations.
+func prepareDownloadDestination(
+	ctx context.Context,
+	originalSrc *objurl.ObjectURL,
+	src *objurl.ObjectURL,
+	dst *objurl.ObjectURL,
+	parents bool,
+) (*objurl.ObjectURL, error) {
+	objname := src.Base()
+	if parents {
+		objname = src.Relative()
+	}
+
+	if originalSrc.HasGlob() {
+		os.MkdirAll(dst.Absolute(), os.ModePerm)
```

Comment:
`os.MkdirAll` returns an error if create dir fails. We should handle it.

Reply:
done.